wordpress-mobile/WordPress-Android
# Android 12: Security and privacy - Pending intents mutability

Closes #16065

References: https://github.com/Automattic/stories-android/pull/721

This PR includes the following changes:

1. Update `PendingIntent` flags to include the mutability flag as part of `Android 12` migration changes for notifications and widgets (see `Details` section). 
2. Upgrade `androidxWorkVersion` to `2.7.0` that fixes a platform bug causing pending intent crash on `targetSdkVersion 31` (2a2f7f0d101ad7633ef65b38f916c3d78892286d, 344179409dfcc47d8f6a0a2e02200668b5b41e0a)
3. Two refactoring commits (faa4288fd2045cc2babb8cb0c70d92359a5b8064, 38f57bbeda1ff2bd4933b4dd49ab50a77c8d3155) that formats files and removes unnecessary version check at places where pending intent flags were introduced.

### Details:

List of all the [notification types](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/WordPress/src/main/java/org/wordpress/android/push/NotificationType.kt#L3) used in the app and classes setting pending intent mutability for them:

Class | NotificationType | Notes
-- | -- | --
GCMMessageHandler (via `NotificationHelper.handleDefaultPush`)| COMMENT | Supports direct reply actions
  | LIKE |  
  | COMMENT_LIKE |  
  | AUTOMATTCHER |  
  | FOLLOW |  
  | REBLOG |  
  | BADGE_RESET |  
  | NOTE_DELETE |  
  | TEST_NOTE |  
  | ZENDESK |  
  | UNKNOWN_NOTE |  
  | AUTHENTICATION |  
  | GROUP_NOTIFICATION |  
NotificationsProcessingService | ACTIONS_RESULT |  
  | ACTIONS_PROGRESS |  
NotificationsPendingDraftsReceiver, PendingDraftsNotificationsUtils (draft posts) | PENDING_DRAFTS | Uses `AlarmManager`
QuickStartReminderReceiver, QuickStartUtils | QUICK_START_REMINDER | Uses `AlarmManager`
PostUploadNotifier | POST_UPLOAD_SUCCESS |  
  | POST_UPLOAD_ERROR |  
  | MEDIA_UPLOAD_SUCCESS |  
  | MEDIA_UPLOAD_ERROR |  
PublishSettingsFragment (scheduled posts) | POST_PUBLISHED | Uses `AlarmManager`
FrameSaveNotifier (storis lib)| STORY_SAVE_SUCCESS |  
  | STORY_SAVE_ERROR |  
  | STORY_FRAME_SAVE_SUCCESS |  
  | STORY_FRAME_SAVE_ERROR |  
ReminderNotifier | BLOGGING_REMINDERS | Uses `WorkManager`
LocalNotificationWorker | CREATE_SITE | Uses `WorkManager`
  | WEEKLY_ROUNDUP | Uses `WorkManager`

Additionally, `WidgetUtils` sets pending intent mutability for `Stats` widgets.

It is important that pending intents mutability for all notifications and widgets are correctly set. Quoting [the doc](https://developer.android.com/guide/components/intents-filters#DeclareMutabilityPendingIntent):
> In most cases, your app should create immutable `PendingIntent` objects,....
However, certain use cases require mutable `PendingIntent` objects instead:
> - Supporting direct reply actions in notifications.
....
> - Scheduling alarms using `AlarmManager`. The mutable `PendingIntent` object allows the system to add the `EXTRA_ALARM_COUNT` intent extra.  This extra represents the number of times that a repeating alarm has been triggered. By containing this extra, the intent can accurately notify an app as to whether a repeating alarm was triggered multiple times, such as when the device was asleep.

For the reasons above, all pending intents are made immutable (8176ba389d6eca228b62119e417933ad679db030) except for comments notifications that allow direct replies and voice remote input (76d8c36b89d5c4f2ab8484d436f5198a0d2ef876).

For the notifications using `AlarmManager` (`PENDING_DRAFTS`, `QUICK_START_REMINDER`, `POST_PUBLISHED`), it seems we do not need a mutable pending intent as they're not repeating alarms.

### To test:

Make sure that at least the special notifications have their pending intent mutability set correctly.

#### Test.1 Comments

1. Login using a wpcom account having sites on your device 
2. Comment on a site post using another account from Calypso 
3. Wait for the comment notification to arrive on the device
5. Expand the notification and direct reply to the comment from the notification
6. Notice that the comment is posted properly

<img height=320 src="https://user-images.githubusercontent.com/1405144/157816907-73d2447a-689c-4446-85cf-100350213d72.png"/>
 

#### Test.2 Quick Start Reminders

1. Fresh launch the app
2. Select a site from the epilogue view
3. When the Quick Start Prompt is shown, tap the "Show me around" option
4. Tap the Customize your site row to launch the QS dialog
5. Tap on Check your site title (the dialog will close and you'll be brought back to My Site)
7. Tap on the blue flashing icon on site title and change the text
8. Notifications are set for two days in the future, so you have to change the date/time on the device for testing
9. Navigate out of the app to the Settings area of your device and locate the area that manages the time.
10. Disable the "use network-provided time"
11. Manually set the date/time for 2 days and 2 minutes from now
12. NOTE: The notification should arrive at the device within a couple of minutes
13. After the notification is shown - don't forget to enable "use network-provided time"

#### Test.3 Post Published

1. Launch the app
2. Click `FAB` to create a post to schedule it for a future time
3. Go to `Post Settings`
4. Go to `Publish Date`
5. Set `Publish Date` two minutes in the future
6. Set `Notification` to `when published`
7. Navigate out of the app
8. The notification should arrive at the device within a couple of minutes

#### Test.4 Adhoc Notification

If you suspect any other notification (from the notifications table above) should have a pending intent mutability other than what is set, test it out or test for a random notification type.

Note that `PENDING_DRAFTS` notifications are not displayed due to an existing bug: https://github.com/wordpress-mobile/WordPress-Android/issues/14240

#### Test.5 Stats Widgets

1. Go to your Home and try to add a Widget
2. Scroll to the `WordPress` 
3. Add Stats widget (you'll need to select a site as an intermediate step)
4. Interact with the widget
5. Make sure everything works as expected

## Merge Instructions
- ~~Make sure that referenced `Stories Lib` PR is merged to `trunk`~~
- ~~Update `Stories Lib` version from `trunk`~~
- ~~Remove `Not Ready for Merge` label~~
- ~~Merge the PR~~

## Regression Notes
1. Potential unintended areas of impact
Push notifications

2. What I did to test those areas of impact (or what existing automated tests I relied on)
See `To Test` section

4. What automated tests I added (or what prevented me from doing so) - N/A

PR submission checklist:
